src/west/app/project.py

```diff
@@ -1316,19 +1316,32 @@ def _rev_type(project, rev=None):
     else:
         return 'other'
 
-def _fetch(project, rev=None):
-    # Fetches rev (or project.revision) from project.url in a way that
+def _fetch(project):
+    # Fetches project.revision from project.url in a way that
     # guarantees any branch, tag, or SHA (that's reachable from a
     # branch or a tag) available on project.url is part of what got
     # fetched.
     #
     # Returns a git revision which hopefully can be peeled to the
-    # newly-fetched SHA corresponding to rev. "Hopefully" because
-    # there are many ways to spell a revision, and they haven't all
-    # been extensively tested.
+    # newly-fetched SHA corresponding to project.revision.
+    #
+    # "Hopefully" because there are many ways to spell a revision, and
+    # they haven't all been extensively tested.
 
-    if not rev:
-        rev = project.revision
+    rev = project.revision
+
+    def fetch_helper(refspec):
+        # -f is needed to avoid errors in case multiple remotes are
+        # present, at least one of which contains refs that can't be
+        # fast-forwarded to our local ref space.
+        #
+        # --tags is required to get tags, since the remote is
+        # specified as a URL. We hold on to tags to avoid losing track
+        # of revisions that got left behind by branches that rebase.
+        project.git(['fetch', '-f', '--tags'] +
```

Comment:
`--tags` rings a few alarm bells.
1a. Fetching a SHA directly had the promise of a huge fetch optimization for repos with many tags (and branches). Gone?
1b. git tags are somewhat flawed because unlike branches they are all global, no namespace. That's why `--tags` is NOT the default, avoids large pollution. Gone? No QUAL_REFS_WEST or anything?
2. Does this solve a real problem? A SHA checked out (HEAD) will never be garbage-collected. So the only problem is when the manifest moves away from this SHA and returns to it (why?) _after_ `gc.reflogExpire` which defaults to 90 days. Then it will merely be fetched again, not a big deal. 
3a If network usage issue 2. is really a problem for someone (for instance because they have a crazy short `gc.reflogExpire`), they can "solve" that issue  in 2 seconds by merely slapping a local branch or local tag onto that commit
3b. Something like `revision-in` / #344 would also avoid this issue.

Reply:
> `--tags` rings a few alarm bells.

Yep. This is the crux of this PR, thanks for finding it right away :smile: .

Context: 

- Nordic's SDK rebases zephyr whenever we cut a new release, to keep a linear history of downstream patches, but allowing `git merge` of upstream zephyr during development.

- We've historically relied on --tags being here to make sure that jumping around between versions during development never creates the possibility for garbage collection of old code, because we tag our history right before we rebase. Running `west update` gets the tag too, so nobody can lose anything.

However, I would be open to dropping --tags here if we can figure out a way to avoid issues.

My main concern is that if I do this:

```
git init
git fetch https://github.com/zephyrproject-rtos/zephyr v2.4.0
git update-ref refs/heads/manifest-rev FETCH_HEAD^{commit}
git checkout --detach manifest-rev
```

Then I have no tags at all, which isn't really what I think we want. I would like to find a solution where fetching a tag gives you a local tag, so you can't lose it later.

Reply:
> That's why `--tags` is NOT the default, avoids large pollution. Gone? No QUAL_REFS_WEST or anything?

```
       By default, any tag that points into the histories being fetched is
       also fetched; the effect is to fetch tags that point at branches that
       you are interested in. 

...
       -n, --no-tags
           By default, tags that point at objects that are downloaded from the
           remote repository are fetched and stored locally.
```

I'm curious what you mean by "NOT the default".

Reply:
We have a comment indicating that the default behavior changes depending on whether you're fetching from a URL or a remote name. 

That appears consistent with what I'm getting from git 2.30.0:

```
$ git remote add origin https://github.com/zephyrproject-rtos/zephyr
$ git fetch origin v2.4.0
From https://github.com/zephyrproject-rtos/zephyr
 * tag                     v2.4.0     -> FETCH_HEAD
```

However, "great, just use a remote" is not a straightforward answer, because we have a contract with users that they can rename their remotes whatever they want, at least so far.

Reply:
> never creates the possibility for garbage collection of old code,

Can you elaborate why this is a real issue?
From a performance point of view `west update` would just fetch again and considering the rebase it would have very little to fetch again because most objects are unchanged by a rebase.
Is the (former) lack of `uploadpack.allow*SHA1InWant` the concern? Either the fallback here or the `revision-in` / #344 would solve this. Something has to anyway.


> I'm curious what you mean by [--tags is] "NOT the default".

I'm merely reading `git help fetch`.

> what I'm getting from git 2.30.0. I still don't get a v2.4.0 tag locally.

I hope there is a solution that does not depend on obscure and changing git implementation details. At least not functionally - performance is of course a different story.

I would not mind some kind of `--tags` _option_ in the west user interface to support this use case, maybe even turned on by default. What bothers me is hardcoding it at the lowest/deepest level in the code which precludes any "high performance" mode.

Reply:
> I'm merely reading `git help fetch`.

What I pasted above is from the same page, and seems to contradict your statement in some ways.

Reply:
I don't think `--no-tags` is the opposite of `--tags`. The default seems to be somewhere in the middle.

Reply:
> From a performance point of view `west update` would just fetch again

I'm thinking of things like long airplane rides without network connectivity, where an ill-timed gc is a showstopper until you land and have wifi again.

> I hope there is a solution that does not depend on obscure and changing git implementation details

I'm not sure git has changed. Just wanted to share the exact version I was using in case you repro and get a difference.

I hope so too.

Reply:
> I don't think `--no-tags` is the opposite of `--tags`. The default seems to be somewhere in the middle.

OK, I get you now, and it seems like you're right.

Lol, that is just classic git.

"We have --foo and --no-foo options."
"What's the default?"
"Neither"

Reply:
I find useful to always keep in mind a "decoder ring" for git's user interface. For instance every time I see "git rebase" I mentally translate to "git replay" - cause I'm changing base less often than not. I just added `--tags` -> `--all-tags` to my internal decoder ring.

It's far from a silver bullet though, "checkout" and "reset" are lost causes https://git-scm.com/book/en/v2/Git-Tools-Reset-Demystified